Error Handling Concept #125 (#131)
exercism/cairo: concepts/error-handling/about.md

            Ok(value) => println!("Division successful: {}", value),
            Err(e) => println!("{}", e),
    }
}
```

## Error with `panic` 
When program encounters an error which is unrecoverable then a `panic` occurs. This will stop the execution, which is helpful in the case where continuing the execution wouldn't make any sense.

+ Example
```rust
fn main() {
    let mut data = array![1, 2];

    if true {
        panic(data);
    }
    println!("This line will not be reached.");
}
```

> `panic!` macro is panic error that allows you to use a string as error message (can include longer than 31 bytes) accepts ByteArrays as argument. 
```rust
panic!("the error for panic! macro is not limited to 31 characters anymore");
```

## `nopanic` Notation
Cairo `nopanic` notation indicates that a function will never panic. It makes code more user-friendly. 
> `nopanic` function can be called only in a function annotated as `nopanic`.

+ Example
```rust
fn function_never_panic() -> felt252 nopanic {
    108
}
```

## `panic_with` Attribute
Cairo `panic_with` attribute simplifies error handling by allowing a function to automatically panic if it returns `None` or `Err`.

+ Example

> ```ruby

Reviewer: ```suggestion
```rust
```

```
enum Result<T, E> {
    Ok: T,
    Err: E,
}
```

> Let's Take an example of `Devide By Zero` 
```
fn divide(a: felt, b: felt) -> Result<felt, &'static str> {
    if b == 0 {
        return Err("Error: Division by zero"); // Return an error if division by zero is attempted
    }
    Ok(a / b)
}
```

```
// Test with division by zero
match divide(10, 0) {
    Ok(result) => println!("Division successful: {}", result),
    Err(e) => println!("{}", e),
}
```
## The `?` Operator
`?` operator is used for minimilistic(concise) errors. Allow to write clean and readable code without explicit values.

>Example
```
fn safe_divide(a: felt, b: felt) -> Result<felt, &'static str> {
    // The ? operator automatically generate the error if it occurs
    let result = divide(a, b)?;
    Ok(result)
}
```

```
match safe_divide(10, 0) {
        Ok(value) => println!("Division successful: {}", value),
        Err(e) => println!("{}", e),
    }
```

## Error with `panic!` 
When program encounters an error which is unrecoverable then a `panic!` occures. This will stop execution and helpful for the case where continuing execution doesn't make any sense.

> Example
```

> fn main() {

Reviewer: Use this example to show the use of a plain `panic`, and move this example to the end of the section so that students still see the macro as well. Add a brief sentence explaining the differences between the two.

    if true {
        panic!("2");
    }
    println!("This line isn't reached");
}
```
> **Output** 
``` 
Run panicked with [1997209042069643135709344952807065910992472029923670688473712229447419591075, 0 (''), 50 ('2'), 1, ].
```

After the panic, the code execution will stop.

## `nopanic` Notation

> Cairo nopanic Notation indicates that function will never panic. It makes code more user-friendly. 

Reviewer: ```suggestion
Cairo `nopanic` notation indicates that a function will never panic. It makes code more user-friendly. 
```

> > `nopanic` function can be called only in a function annotated as `nopanic`.

Reviewer: This is a great use of a note block :+1:

> Example
```
fn function_never_panic() -> felt252 nopanic {

>     println!("No panic");

Reviewer: This code does not compile.
Reuse the example from https://book.cairo-lang.org/ch09-01-unrecoverable-errors-with-panic.html#nopanic-notation

}
```

## `panic_with` Attribute
Cairo `panic_with` attribute simplifies error handling by allowing a function to automatically panic if it returns `None` or `Err`.

> Example
```

> fn check_not_zero(value: u128) -> Option<u128> {

Reviewer: You forgot to add the `panic_with` attribute :)

    if value == 0 {
        Option::None
    } else {
        Option::Some(value)
    }
}

fn main() {
    check_not_zero(0); // -> returns None
    wrap_not_zero(0);  // -> panics with 'value is 0'
}
```

---
## Summary

> Error handling involves identifying and addressing errors to execute a program smoothly. `Recoverable errors` doesn't stop the execution of program whereas `Unrecoverable Errors with panic` results runtime errors. Read more about `Error Handling` in [Cairo Book](https://book.cairo-lang.org/ch09-00-error-handling.html)

Reviewer: Remove the summary, there's generally no need for it as the concepts are themselves quite short

    let input_to_u8: u8 = parse_u8(input)?;
    // DO SOMETHING
    let res = input_to_u8 - 1;
    Result::Ok(res)
}

fn main() {
    let number: felt252 = 258;
    match do_something_with_parse_u8(number) {
        Result::Ok(value) => println!("Result: {}", value),
        Result::Err(e) => println!("Error: {}", e),
    }
}
```

## Error with `panic`

When the program encounters an error that is unrecoverable, then a `panic` occurs. This will stop the execution, which is helpful in the case where continuing the execution wouldn't make any sense.

```rust
fn main() {
    let data = array![1, 2];

    if true {
        panic(data);
    }
    println!("This line will not be reached.");
}
```

Running this example would print this in the console:

```console
Run panicked with [1, 2, ].
```

There is also a special macro, called `panic!` that allows you to use a `ByteArray` as the error message.

```rust
panic!("The error for the panic! Error message is not limited to 31 characters anymore");
```

## `nopanic` Notation

Cairo `nopanic` notation indicates that a function will never panic. It makes code more user-friendly.

> > `nopanic` function can be called only in a function annotated as `nopanic`.

Reviewer: ```suggestion

> `nopanic` function can be called only in a function annotated as `nopanic`.
```

> 

Reviewer: ```suggestion
```

`?` operator is used for concise and implicit error handling, letting the calling function deal with any errors that might occur.

Example
```rust
fn safe_divide(a: u32, b: u32) -> Result<felt252, ByteArray> {
    // The ? operator automatically generate the error if it occurs
    let result = divide(a, b)?;
    Result::Ok(result)
}

fn main() {
    match safe_divide(10, 0) {
        Ok(value) => println!("Division successful: {}", value),
        Err(e) => println!("{}", e),
    }
}
```

## Error with `panic` 
When program encounters an error which is unrecoverable then a `panic` occurs. This will stop the execution, which is helpful in the case where continuing the execution wouldn't make any sense.

Example
```rust
fn main() {
    let data = array![1, 2];

    if true {
        panic(data);
    }
    println!("This line will not be reached.");
}
```
Running this example would print this in the console:

```console
Run panicked with [1, 2, ].
```

There is also a special macro called `panic!`, that allows you to use a `ByteArray` as the error message.
```rust
panic!("The error for the panic! Error message is not limited to 31 characters anymore");
```

## `nopanic` Notation

> Cairo `nopanic` notation indicates that a function will never panic. It makes code more user-friendly. 

Reviewer: Don't forget to [run the `./bin/lint_markdown.sh` command in you terminal](https://github.com/exercism/cairo/pull/131#discussion_r1730300530)

```rust
fn divide(a: felt252, b: felt252) -> Result<felt252, ByteArray> {
    if b == 0 {
        return Err("Error: Division by zero"); // Return an error if division by zero is attempted
    }
    Ok(a / b)
}

fn main() {
    // Test with division by zero
    match divide(10, 0) {
        Ok(result) => println!("Division successful: {}", result),
        Err(e) => println!("{}", e),
    }
}
```
## The `?` Operator
`?` operator is used for concise and implicit error handling, letting the calling function deal with any errors that might occur.

+ Example
```rust
fn safe_divide(a: felt252, b: felt252) -> Result<felt252, ByteArray> {
    // The ? operator automatically generate the error if it occurs
    let result = divide(a, b)?;
    Ok(result)
}

fn main() {
    match safe_divide(10, 0) {
            Ok(value) => println!("Division successful: {}", value),
            Err(e) => println!("{}", e),
    }
}
```

## Error with `panic` 
When program encounters an error which is unrecoverable then a `panic` occurs. This will stop the execution, which is helpful in the case where continuing the execution wouldn't make any sense.

+ Example
```rust
fn main() {

>     let mut data = array![1, 2];

Reviewer: ```suggestion
    let data = array![1, 2];
```
nit: no need for `data` to be mutable

fn safe_divide(a: felt252, b: felt252) -> Result<felt252, ByteArray> {
    // The ? operator automatically generate the error if it occurs
    let result = divide(a, b)?;
    Ok(result)
}
```

```
match safe_divide(10, 0) {
        Ok(value) => println!("Division successful: {}", value),
        Err(e) => println!("{}", e),
fn main() {
    match safe_divide(10, 0) {
            Ok(value) => println!("Division successful: {}", value),
            Err(e) => println!("{}", e),
    }
}
```

## Error with `panic!` 
When program encounters an error which is unrecoverable then a `panic!` occures. This will stop execution and helpful for the case where continuing execution doesn't make any sense.
## Error with `panic` 
When program encounters an error which is unrecoverable then a `panic` occurs. This will stop the execution, which is helpful in the case where continuing the execution wouldn't make any sense.

> Example
```
+ Example
```rust
fn main() {
    let mut data = array![1, 2];

    if true {
        panic!("2");
        panic(data);
    }
    println!("This line isn't reached");
    println!("This line will not be reached.");
}
```

> > **Output** 

Reviewer: Hmm, maybe add a line like:
``````markdown
Running this example would print this in the console:

```console
Run panicked with [1, 2, ].
```

> > `panic!` macro is panic error that allows you to use a string as error message (can include longer than 31 bytes) accepts ByteArrays as argument. 

Reviewer: ```suggestion
There is also a special macro called `panic!`, that allows you to use a `ByteArray` as the error message.
 
```
No need to make this a note, it's really a part of the concept

> panic!("the error for panic! macro is not limited to 31 characters anymore");

Reviewer: ```suggestion
panic!("The error for the panic! Error message is not limited to 31 characters anymore");
```
nit

> Cairo `nopanic` notation indicates that a function will never panic. It makes code more user-friendly. 

Reviewer: You will need to format this file according to markdown formatting rules.

To see potential issues, open your terminal, navigate to the root of this repository and run:
```bash
./bin/lint_markdown.sh 
```
